corehq/apps/analytics/tests/test_partner_analytics_utils.py

```diff
@@ -0,0 +1,277 @@
+import calendar
+import datetime
+
+from unittest import mock
+
+from django.test import TestCase
+
+from corehq.apps.accounting.models import DomainUserHistory
+from corehq.apps.analytics.models import (
+    PartnerAnalyticsContact,
+    PartnerAnalyticsReport,
+    PartnerAnalyticsDataPoint,
+)
+from corehq.apps.analytics.utils.partner_analytics import (
+    get_number_of_mobile_workers,
+    generate_monthly_mobile_worker_statistics,
+    track_partner_access,
+    generate_monthly_web_user_statistics,
+    generate_monthly_submissions_statistics,
+    get_csv_details_for_partner,
+    send_partner_emails,
+    NUMBER_OF_MOBILE_WORKERS,
+    NUMBER_OF_WEB_USERS,
+    NUMBER_OF_SUBMISSIONS,
+    ACCESS_ODATA,
+)
+from corehq.apps.domain.models import Domain
+from corehq.apps.es.tests.utils import es_test
+from dimagi.utils.dates import get_start_and_end_dates_of_month
+
+
+def _get_fake_number_of_mobile_workers(domain, _year, _month):
+    return {
+        'test-001': 10,
+        'test-002': 20,
+        'test-003': 30,
+    }[domain]
+
+
+def _get_fake_number_of_web_users(domain, _year, _month):
+    return {
+        'test-001': 11,
+        'test-002': 21,
+        'test-003': 31,
+    }[domain]
+
+
+def _get_fake_number_of_submissions(domain, _year, _month):
+    return {
+        'test-001': 11,
+        'test-002': 22,
+        'test-003': 33,
+    }[domain]
+
+
+@es_test
+class TestPartnerAnalyticsDataUtils(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls.date_start, cls.date_end = get_start_and_end_dates_of_month(2021, 11)
+        cls.domain = Domain.get_or_create_with_name('test-partner-analytics', is_active=True)
+
+        ## Data for Mobile Workers Tests
+        # test month
+        DomainUserHistory.objects.create(
+            domain=cls.domain.name,
+            record_date=cls.date_end,
+            num_users=3
+        )
+        # previous month
+        DomainUserHistory.objects.create(
+            domain=cls.domain.name,
+            record_date=cls.date_start - datetime.timedelta(days=1),
+            num_users=6
+        )
+        # next month
+        DomainUserHistory.objects.create(
+            domain=cls.domain.name,
+            record_date=cls.date_end + datetime.timedelta(days=1),
+            num_users=9
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        DomainUserHistory.objects.all().delete()
+        cls.domain.delete()
+        super().tearDownClass()
+
+    def test_get_number_of_mobile_workers(self):
+        self.assertEqual(
+            get_number_of_mobile_workers(
+                self.domain.name, self.date_start.year, self.date_start.month
+            ),
+            3
+        )
+
+
+class TestPartnerAnalyticsReportUtils(TestCase):
```

Comment:
Is the `corehq.apps.es.fake` module deprecated or just not widely known?  Reading [the docs](https://commcare-hq.readthedocs.io/es_fake.html) it looks like it would be a good fit for your needs here, but I've never used it myself.  If you used it here, then the `es_test` decorator wouldn't be relevant at all.

Reply:
Alternatively, the _full_ elastic boilerplate to populate real, queryable docs is quite verbose and might look something like:
```python
from corehq.elastic import get_es_new, send_to_elasticsearch
from corehq.pillows.mappings import USER_INDEX_INFO, XFORM_INDEX_INFO
from corehq.pillows.user import transform_user_for_elasticsearch
from corehq.pillows.xform import transform_xform_for_elasticsearch

# <snip>

    def setUpClass(cls):

        # <snip>

        cls.es = get_es_new()
        initialize_index_and_mapping(cls.es, USER_INDEX_INFO)
        initialize_index_and_mapping(cls.es, XFORM_INDEX_INFO)
        # init Elastic forms
        cls.forms = [
            ...
        ]
        for form in cls.forms:
            elastic_form = transform_xform_for_elasticsearch(form.to_json())
            send_to_elasticsearch('forms', elastic_form)
        # init Elastic users
        cls.users = [
            ...
        ]
        for user in cls.users:
            elastic_user = transform_user_for_elasticsearch(user.to_json())
            send_to_elasticsearch('users', elastic_user)
        cls.es.indices.refresh(XFORM_INDEX_INFO.alias)
        cls.es.indices.refresh(USER_INDEX_INFO.alias)

    def tearDownClass(cls):
        ensure_index_deleted(USER_INDEX_INFO.alias)
        ensure_index_deleted(XFORM_INDEX_INFO.alias)

        # <snip>
```